app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedFragmentTest.kt

```diff
@@ -0,0 +1,565 @@
+import org.oppia.android.domain.oppialogger.loguploader.WorkManagerConfigurationModule
+import org.oppia.android.domain.platformparameter.PlatformParameterModule
+import org.oppia.android.domain.question.QuestionModule
+import org.oppia.android.domain.topic.PrimeTopicAssetsControllerModule
+import org.oppia.android.testing.AccessibilityTestRule
+import org.oppia.android.testing.TestLogReportingModule
+import org.oppia.android.testing.robolectric.RobolectricModule
+import org.oppia.android.testing.story.StoryProgressTestHelper
+import org.oppia.android.testing.threading.TestCoroutineDispatchers
+import org.oppia.android.testing.threading.TestDispatcherModule
+import org.oppia.android.testing.time.FakeOppiaClock
+import org.oppia.android.testing.time.FakeOppiaClockModule
+import org.oppia.android.util.accessibility.AccessibilityTestModule
+import org.oppia.android.util.caching.testing.CachingTestModule
+import org.oppia.android.util.gcsresource.GcsResourceModule
+import org.oppia.android.util.logging.LoggerModule
+import org.oppia.android.util.logging.firebase.FirebaseLogUploaderModule
+import org.oppia.android.util.parser.html.HtmlParserEntityTypeModule
+import org.oppia.android.util.parser.image.GlideImageLoaderModule
+import org.oppia.android.util.parser.image.ImageParsingModule
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Tests for [MarkTopicsCompletedFragment]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(
+  application = MarkTopicsCompletedFragmentTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+class MarkTopicsCompletedFragmentTest {
+
+  private val internalProfileId = 0
+  private lateinit var profileId: ProfileId
+
+  @Inject
+  lateinit var storyProgressTestHelper: StoryProgressTestHelper
+
+  @Inject
+  lateinit var fakeOppiaClock: FakeOppiaClock
+
+  @Inject
+  lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+
+  @Inject
+  lateinit var context: Context
+
+  @get:Rule
+  val activityTestRule = ActivityTestRule(
+    MarkTopicsCompletedActivity::class.java,
+    /* initialTouchMode= */ true,
+    /* launchActivity= */ false
+  )
+
+  @get:Rule
+  val accessibilityTestRule = AccessibilityTestRule()
+
+  @Before
+  fun setUp() {
+    profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    setUpTestApplicationComponent()
+    fakeOppiaClock.setFakeTimeMode(FakeOppiaClock.FakeTimeMode.MODE_UPTIME_MILLIS)
+    testCoroutineDispatchers.registerIdlingResource()
+  }
+
+  @After
+  fun tearDown() {
+    testCoroutineDispatchers.unregisterIdlingResource()
+  }
+
+  private fun setUpTestApplicationComponent() {
+    ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
+  }
+
+  @Test
+  fun testMarkTopicsCompletedFragment_topicsAreShown() {
+    launch<MarkTopicsCompletedActivity>(
```

Comment:
If this test is launching MarkTopicsCompletedActivity then it's testing the activity, not the fragment. To test the fragment we should host it in a test activity on its own.

See https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt & https://github.com/oppia/oppia-android/blob/develop/app/src/main/java/org/oppia/android/app/player/state/testing/StateFragmentTestActivity.kt for an idea on how this works.

Reply:
I had a look at these and some other test activities and their corresponding prod files and I am a bit confused. 

Actually, I have noticed that test activities are there for cases where the fragment does not have its own activity, for example: `StateFragment` does not have a `StateActivity`.

Also, in some cases, we are launching the prod activity in `FragmentTest`, for example: `OnboardingFragmentTest` launches `OnboardingActivity` in its tests. There might be some reason for it that I may have missed.

Another case is where we have test activities even when the prod activities exist. Now if I'm to create a test activity in this case, I'm not exactly sure what functionalities should be there in this test activity. Also, in some of the cases, the test activity uses the prod activity presenter and in some cases, we have also created a test activity presenter. So, I'm also confused here.

Please help me understand this and guide me on how should I proceed.

Reply:
These are good questions @yashraj-01. In general, components should always be tested in isolation (meaning that for any fragment it should always be tested via a test activity that hosts it). We aren't consistent on this everywhere today, and there are existing examples you pointed out that could also benefit from this change.

Regarding what the test activity should do: essentially the bare minimum to be able to test the fragment. We actually want to keep them as simple as possible, and often avoid introducing presenters altogether and just have the test activity.

Does this help clarify?

Reply:
Thanks @BenHenning for the clarification. I have introduced `MarkTopicsCompletedTestActivity` from what I have understood from your comments. PTAL and confirm if it is in the right direction or not? Thanks.

Reply:
Yep, this exactly addresses my comment. Thanks!